### DISCUSSION TITLE
`take` not working as expected on relation query

### DISCUSSION THREAD

**Discussion by ItsArnob** [prisma/prisma#13843] [Q&A] [answered]
Hello! This is my first time starting a discussion on github and interacting with a community in general, so I apologize in advance if i've made any mistakes 😅
  
So, I'm having an issue with a query, it seems like `take` is not working as expected on the `messages` field. I think its downloading all the records and then returning 3 records, atleast that's what i understood from the query log.  

this is the query I'm executing:
```ts
await prisma.chat.findUnique({
    where: {
        id: "62a9cdf54be1b9e4d35d9d8b"
    },
    include: {
        messages: {
            take: 3
        }
    }        
})
```
this query takes a long time (5.3 minutes?? on my decent-ish internet)  to complete since the `messages` collection has 100K records...
here are the queries that prisma generated:
``` 
[api] prisma:query  db.chats.aggregate([ { $match: { $expr: { $and: [ { $and: [ { $eq: [ "$_id", ObjectId("62a9cdf54be1b9e4d35d9d8b"), ], }, { $or: [ { $ne: [ { $ifNull: [ "$_id", null, ], }, null, ], }, { $eq: [ "$_id", null, ], }, ], }, ], }, ], }, }, }, { $project: { _id: 1, name: 1, chatType: 1, createdAt: 1, recipients.userId: 1, recipients.nickname: 1, lastMessageId: 1, }, }, ])
```
```
[api] prisma:query db.messages.aggregate([ { $match: { $expr: { $and: [ { $in: [ "$chatId", [ "62a9cdf54be1b9e4d35d9d8b", ], ], }, { $or: [ { $ne: [ { $ifNull: [ "$chatId", null, ], }, null, ], }, { $eq: [ "$chatId", null, ], }, ], }, ], }, }, }, { $sort: { _id: 1, }, }, { $project: { _id: 1, chatId: 1, authorId: 1, content: 1, deleted: 1, createdAt: 1, }, }, ])
```
And here's my schema: 
```prisma
model Chat {

  id            String          @id @default(auto()) @map("_id") @db.ObjectId
  name          String?
  chatType      ChatType
  createdAt     DateTime        @default(now())
  recipients    chatRecipient[]
  messages      Message[]       @relation("messages")
  // lastMessage   Message?        @relation("lastMessage", fields: [lastMessageId], references: [id], onDelete: NoAction, onUpdate: NoAction)
  lastMessageId String?         @unique

  @@map("chats")
}

model Message {
  id              String   @id @default(auto()) @map("_id") @db.ObjectId
  chat            Chat     @relation("messages", fields: [chatId], references: [id])
  // lastMessageChat Chat?    @relation("lastMessage")
  chatId          String
  authorId        String
  content         String?
  deleted         Boolean?
  createdAt       DateTime @default(now())

  @@map("messages")
}
```
The aggregation pipeline for the `messages` collection doesn't include `$limit` so I think mongodb returns all the documents. I don't really know if the problem lies within my schema or the query or its just a bug so any help is appreciated 😃 

I might just be making a silly mistake, this is my first time experimenting with prisma

**Comment by austincrim** [comment, marked as answer]
Hey @ItsArnob 👋,

This definitely seems unexpected! I saw the same results that you shared, but only when doing `take` with a nested `include` on `findUnique`. Using `take` at the top level of a `findMany` seems to use the `$limit` operator.

Would you mind filling out a [bug report](https://github.com/prisma/prisma/issues/new?assignees=&labels=kind%2Fbug&template=bug_report.yml)?

In the meantime, you could use our [`$aggregateRaw` API](https://www.prisma.io/docs/concepts/components/prisma-client/raw-database-access#aggregateraw) to get something more performant 🙂.

**Reply by ItsArnob** [reply]
thanks for your reply!
I tried using `aggregateRaw` and its much faster (talk about ~700ms!) now.  I think this might be related to #10949 with SQL databases.
and yeah sure, I've just filed a bug report here! #13865  🙂

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
